File: Array/LCISFinder/go/LCISFinder.go
package _go

// typically check the continuos increasing subarray and store the max in a third var
// Time complexity: O(n)
// Space complexity: O(1)
func findLengthOfLCIS(nums []int) int {
	if len(nums) <= 1 {
		return len(nums)
	}
	ans, count := 1, 1
	for i := 0; i < len(nums)-1; i++ {
		if nums[i] < nums[i+1] {
			count++
			ans = max(ans, count)
		} else {
			count = 1
		}
	}
	return ans
}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// just used an extra storage to store all those states
// Time complexity: O(n)
// Space complexity: O(n)
func findLengthOfLCIS2(nums []int) int {
	if len(nums) <= 1 {
		return len(nums)
	}
	dp := make([]int, len(nums))
	dp[0] = 1
	ans := 1
	for i := 1; i < len(nums); i++ {
		if nums[i] > nums[i-1] {
			dp[i] = dp[i-1] + 1
		} else {
			dp[i] = 1
		}
		ans = max(ans, dp[i])
	}
	return ans
}

File: Array/LCISFinder/cpp/LCISFinder.cpp
// Problem: https://leetcode.com/problems/longest-continuous-increasing-subsequence/

#include <vector>

using namespace std;

class LCISFinder {
public:
    int findLengthOfLCIS(vector<int>& nums) {
        int len = 1;
        int max_len = 1;
        for (int idx = 0; idx < (int)(nums.size() - 1); ++idx) {
            if (nums[idx] < nums[idx + 1]) { ++len; max_len = max(max_len, len); }
            else len = 1;
        }
        return max_len;
    }
};